src/lsm/zig_zag_merge.zig

```diff
@@ -0,0 +1,665 @@
+                        probe_key = key;
+
+                        // Setting all previous streams as `true` except the drained ones.
+                        probing.setRangeValue(.{ .start = 0, .end = stream_index }, true);
+                        probing.setIntersection(drained.complement());
+                    } else if (switch (it.direction) {
+                        .ascending => key < probe_key,
+                        .descending => key > probe_key,
+                    }) {
+                        // The stream is behind.
+                        probing.set(stream_index);
+                    } else {
+                        // The key matches.
+                        assert(key == probe_key);
+                    }
+                }
+
+                // Probing the buffered streams that did not match the key.
+                var probing_iterator = probing.iterator(.{ .kind = .set });
+                while (probing_iterator.next()) |stream_index| {
+                    stream_probe(it.context, @intCast(stream_index), probe_key);
+
+                    const key = stream_peek(it.context, @intCast(stream_index)) catch |err| {
+                        switch (err) {
+                            error.Empty => return null,
+                            error.Drained => {
+                                drained.set(stream_index);
+                                probing.unset(stream_index);
+                                continue;
+                            },
+                        }
+                    };
+
+                    if (key == probe_key) {
+                        probing.unset(stream_index);
+                    } else {
+                        assert(switch (it.direction) {
+                            .ascending => key > probe_key,
+                            .descending => key < probe_key,
+                        });
+                    }
+                }
+            }
+
+            if (drained.count() == it.streams_count) {
+                //Can't probe if all streams are drained.
+                assert(probe_key == key_min);
+                return error.Drained;
+            }
+
+            assert(probe_key != key_min);
+
+            // At this point, all the buffered streams have produced a match
+            // that will be used to probe the drained streams.
+            if (drained.count() > 0) {
+                var drained_iterator = drained.iterator(.{ .kind = .set });
+                while (drained_iterator.next()) |stream_index| {
+                    stream_probe(it.context, @intCast(stream_index), probe_key);
+
+                    // The stream must remain drained after probed.
+                    assert(stream_peek(it.context, @intCast(stream_index)) == error.Drained);
+                }
+                return error.Drained;
+            }
+
+            return probe_key;
+        }
+
+        fn peek_key_classic(it: *ZigZagMergeIterator) error{Drained}!?Key {
```

Comment:
`peek_keys_classic` is the reference implementation of the ZigZag over 2 streams (the "seed" and the "probe") .
This implementation is less efficient since it can unnecessarily probe streams with a seed that may be behind the other streams not included in the current iteration. For example, it zig-zags only between `A` and `B` first and then uses the matching key as the seed to zig-zag into `C`, only to discover that `C` might be far ahead from the seed, restarting the process with `C` and `A` .

The optimized  `peek_keys` version, reads all streams in one pass, probing those that don't match at once.
For example, it peeks at `A`,`B`, and `C` for the most advanced key. If `C` produces the key, it probes `A` and `B` and repeats the process.